README.md
# RXTX, using the serial port from Clojure JVM

[![Clojars
Project](https://img.shields.io/clojars/v/io.helins/rxtx.svg)](https://clojars.org/io.helins/rxtx)

[![Cljdoc](https://cljdoc.org/badge/io.helins/rxtx)](https://cljdoc.org/d/io.helins/rxtx)

Simple API for doing serial IO.

Based on [jRxTx](https://github.com/openmuc/jrxtx).

Supports a (very) wide variety of platforms.


## Installation

There is a bit of a setup but nothing to worry about, it is fairly easy.

For more information about the process or if something goes wrong, go to
[jRxTx](https://github.com/openmuc/jrxtx).

For instance, on a debian based distribution :

```
sudo apt install librxtx-java
```

When starting your application or a repl, check if the "java.library.path"
property contains a path to the installed native libs. On debian, that should be
'/usr/lib/jni' :

```clj
(System/getProperty "java.library.path")
```

If this property is not properly set, you must do it yourself.

In your project file :
```clj
:jvm-opts ["-Djava.library.path=/PATH/TO/LIBS"]
```

When executing your uberjar :
```sh
java -Djava.library.path=/PATH/TO/LIBS -jar your_program.jar
```


## Usage

This is a very brief overview.

The [full API is available on Cljdoc](https://cljdoc.org/d/io.helins/rxtx).

In short, without error handling :

```clj
(require '[helins.rxtx :as rxtx])


(with-open [port (rxtx/serial-port "/dev/ttyUSB0"
                                   {:rxtx/baud-rate 9600
                                    :rxtx/parity    :even})]

  ;; Different things can be written to a serial port besides a byte array.
  ;;
  (rxtx/write port
              "Hello ")
  (rxtx/write port
              \w)
  (rxtx/write port
              111)
  (rxtx/write port
              [114 108 100])

  ;; Reads the answer, up to 16 unsigned bytes, but with a timeout of 2000 milliseconds.
  (println :answer
           (String. (byte-array (rxtx/read port
                                           16
                                           2000)))))
```


## License

Copyright © 2017 Adam Helinski

Licensed under the term of the Mozilla Public License 2.0, see LICENSE.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.



## [Unreleased]

### Added

### Changed

### Fixed



## [2.0.0] - 2021-03-25

### Changed

- Name pluralization scheme
    - available-bytes -> available-byte+
    - available-ports -> available-port+
    - defaults        -> default+
- Namespace parameters under "rxtx"
- Relicensed under MPL 2.0
- Renamed group id from "dvlopt" to "io.helins"

### Fixed

- Keywords for stop-bits which do not begin with a number (illegal)



## [1.0.1] - 2020-01-18

### Fixed

- Improved support for `with-open`



## [1.0.0] - 2018-06-14

### Changed

- New, simpler API



## [0.0.0-alpha1] - 2017-11-04

### Changed

- Improved documentation



## [0.0.0-alpha0] - 2017-09-06

### Added

- First API iteration



[Unreleased]: https://github.com/helins/rxtx.clj/compare/2.0.0...HEAD
[2.0.0]: https://github.com/helins/rxtx.clj/compare/1.0.1...2.0.0
[1.0.1]: https://github.com/helins/rxtx.clj/compare/1.0.0...1.0.1
[1.0.0]: https://github.com/helins/rxtx.clj/compare/0.0.0-alpha1...1.0.0
[0.0.0-alpha1]: https://github.com/helins/rxtx.clj/compare/0.0.0-alpha0...0.0.0-alpha1
[0.0.0-alpha0]: https://github.com/helins/rxtx.clj/tree/0.0.0-alpha0


project.clj
(defproject dvlopt/rxtx
            "1.0.1"

  :description  "https://github.com/dvlopt/rxtx"
  :url          "Serial IO based on RXTX"
  :license      {:name "Eclipse Public License"
                 :url  "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[dvlopt/void       "0.0.0"]
                 [org.openmuc/jrxtx "1.0.1"]]
  :profiles     {:dev     {:source-paths ["dev"]
                           :main         user
                           :dependencies [[criterium              "0.4.4"]
                                          [org.clojure/clojure    "1.9.0"]
                                          [org.clojure/test.check "0.10.0-alpha2"]]
                           :plugins      [[lein-codox      "0.10.3"]
                                          [venantius/ultra "0.5.1"]]
                           :codox        {:output-path  "doc/auto"
                                          :source-paths ["src"]}
                           :global-vars  {*warn-on-reflection* true}}})


src/main/helins/rxtx.clj
;; This Source Code Form is subject to the terms of the Mozilla Public
;; License, v. 2.0. If a copy of the MPL was not distributed with this
;; file, You can obtain one at https://mozilla.org/MPL/2.0/.


(ns helins.rxtx

  "IO on serial ports based on jRXTX.
   
   Any IO operation will throw an exception in case of failure."

  {:author "Adam Helinski"}

  (:import (org.openmuc.jrxtx DataBits
                              FlowControl
                              Parity
                              SerialPort
                              SerialPortBuilder
                              SerialPortException
                              SerialPortTimeoutException
                              StopBits)
           (java.io InputStream
                    OutputStream)
           java.lang.AutoCloseable)
  (:refer-clojure :exclude [flush
                            read]))


;;;;;;;;;; Default values


(def default+

  "Defaut values for keys and options used throughout this namespace."

  {:rxtx/baud-rate    9600
   :rxtx/data-bit+    8
   :rxtx/flow-control :none
   :rxtx/parity       :even
   :rxtx/stop-bits    :1})



(defn- -obtain

  ;; Retrieves a value or relies on default one.

  [k hmap]

  (or (get hmap
           k)
      (get default+
           k)))


;;;;;;;;;; Conversions


(defn- -number->data-bits

  ;; Converts a number to a DataBits enum value.

  ^DataBits

  [^long data-bits]

  (condp identical?
         data-bits
    5 DataBits/DATABITS_5
    6 DataBits/DATABITS_6
    7 DataBits/DATABITS_7
    8 DataBits/DATABITS_8))



(defn- -data-bits->number

  ;; Converts a Databits enum value to an number.

  ^long

  [data-bits]

  (condp identical?
         data-bits
    DataBits/DATABITS_5 5
    DataBits/DATABITS_6 6
    DataBits/DATABITS_7 7
    DataBits/DATABITS_8 8))



(defn- -kw->stop-bits

  ;; Converts a number to a StopBits enum value.

  ^StopBits

  [stop-bits]

  (condp identical?
         stop-bits
    :stop-bits-1   StopBits/STOPBITS_1
    :stop-bits-1.5 StopBits/STOPBITS_1_5
    :stop-bits-2   StopBits/STOPBITS_2))



(defn- -stop-bits->kw

  ;; Converts a StopBits enum value to a number.

  [stop-bits]

  (condp identical?
         stop-bits
    StopBits/STOPBITS_1   :stop-bits-1
    StopBits/STOPBITS_1_5 :stop-bits-1.5
    StopBits/STOPBITS_2   :stop-bits-2))



(defn- -kw->parity

  ;; Converts a kw to a Parity enum value.

  ^Parity

  [kw]

  (condp identical?
         kw
    :even  Parity/EVEN
    :mark  Parity/MARK
    :none  Parity/NONE
    :odd   Parity/ODD
    :space Parity/SPACE))



(defn- -parity->kw

  ;; Converts a Parity enum value to a kw.

  [^Parity parity]

  (condp identical?
         parity
    Parity/EVEN  :even
    Parity/MARK  :mark
    Parity/NONE  :none
    Parity/ODD   :odd
    Parity/SPACE :space))



(defn- -kw->flow-control

  ;; Converts a kw to a FlowControl enum value.

  ^FlowControl

  [kw]

  (condp identical?
         kw
    :none     FlowControl/NONE
    :rts-cts  FlowControl/RTS_CTS
    :xon-xoff FlowControl/XON_XOFF))



(defn- -flow-control->kw

  "Converts a FlowControl enum value to a kw."

  [^FlowControl flow-control]

  (condp identical?
         flow-control
    FlowControl/NONE     :none
    FlowControl/RTS_CTS  :rts-cts
    FlowControl/XON_XOFF :xon-xoff))


;;;;;;;;;; Opening and closing a serial port


(defn available-port+

  "Gets a set of available serial ports.

   Prints an error if a port is already locked but does not throw."

  []

  (into #{}
        (map str)
        (SerialPortBuilder/getSerialPortNames)))



(defn close

  "Closes a serial port."

  [^AutoCloseable port]

  (.close port))



(defn serial-port

  "Opens a serial port at the given path.

   The configuration map may contain:

   | Key | Description |
   |---|---|
   | :rxtx/baud-rate | Preferably a standard baud rate |
   | :rxtx/data-bits | How many bits per data char one of #{5 6 7 8} |
   | :rxtx/flow-control | Handshaking for preventing the sender from sending too much too fast, see below |
   | :rxtx/parity | Error detection, see below |
   | :rxtx/stop-bits | Bits sent at the end of each data char, see below |

   Flow control is one of:

   | Keyword | Description |
   |---|---|
   | :none | No handshaking |
   | :rts-cts | Hardware flow-control |
   | :xon-xoff | Software flow-control |

   Parity is one of:

   | Keyword | Description |
   |---|---|
   | :even | An even parity bit will be sent for each data char |
   | :mark | Mark parity bit |
   | :none | No parity bit |
   | :odd | Odd parity will be sent for each data char |
   | :space |Space parity |

   Stop-bits is one of:

   - :stop-bits-1
   - :stop-bits-1.5
   - :stop-bits-2

   See [[default+]] for default values relative to these options.


   ```clojure
   (serial-port \"/dev/ttyUSB0\"
                {:rxtx/baud-rate 2400
                 :rxtx/parity    :none})
   ```"


  (^AutoCloseable
    
   [path]

   (serial-port path
                nil))


  (^AutoCloseable
    
   [path port-options]

   (.build (doto (SerialPortBuilder/newBuilder path)
             (.setBaudRate (-obtain :rxtx/baud-rate
                                    port-options))
             (.setDataBits (-number->data-bits (-obtain :rxtx/data-bits
                                                        port-options)))
             (.setStopBits (-kw->stop-bits (-obtain :rxtx/stop-bits
                                                    port-options)))
             (.setParity (-kw->parity (-obtain :rxtx/parity
                                               port-options)))
             (.setFlowControl (-kw->flow-control (-obtain :rxtx/flow-control
                                                          port-options)))))))


;;;;;;;;;; Describing a serial port, reconfiguration, and getting streams


(defn describe 

  "Describes the current status of the given serial port by providing a map containing
   the same keys as the option for opening a serial port as well as a :rxtx/name and :rxtx/closed?.

   See [[serial-port]]."

  [^SerialPort port]

  {:rxtx/baud-rate    (.getBaudRate port)
   :rxtx/closed?      (.isClosed port)
   :rxtx/data-bits    (-data-bits->number (.getDataBits port))
   :rxtx/flow-control (-flow-control->kw (.getFlowControl port))
   :rxtx/name         (.getPortName port)
   :rxtx/parity       (-parity->kw (.getParity port))
   :rxtx/stop-bits    (-stop-bits->kw (.getStopBits port))})




(defn input-stream

  "Given an open serial port, retrieves the associated `java.io.InputStream`.
  
   Should not be needed unless really needed."

  ^InputStream

  [^SerialPort port]

  (.getInputStream port))




(defn output-stream

  "Given an open serial port, retrieves the associated `java.io.OutputStream`.

   Should not be needed unless really needed."

  ^OutputStream

  [^SerialPort port]

  (.getOutputStream port))



(defn reconfigure

  "Reconfigures the given serial port on the fly by providing a map containing the same options as for
   opening a new port.

   Missing options will not resolve to default values.

   See [[serial-port]]."

  [^SerialPort port port-options]

  (doseq [[option
           value]  port-options]
    (case option
      :rxtx/baud-rate    (.setBaudRate    port
                                          value)
      :rxtx/data-bits    (.setDataBits    port
                                          (-number->data-bits value))
      :rxtx/flow-control (.setFlowControl port
                                          (-kw->flow-control value))
      :rxtx/parity       (.setParity      port
                                          (-kw->parity value))
      :rxtx/stop-bits    (.setStopBits    port
                                          (-kw->stop-bits value))))
  nil)


;;;;;;;;;; Reading an input stream


(defn available-byte+

  "Estimates how many bytes are available for reading right now."

  [^SerialPort port]

  (.available (input-stream port)))



(defn flush

  "Ensures written bytes are properly flushed."

  [^SerialPort port]

  (.flush (output-stream port)))



(defn read

  "Reads an arbitrary amount of unsigned bytes from the serial port and returns them as a sequence.

   If a timeout is provided, unblocks once it is up and returns what has been read up to this point.
   It not, blocks until `n` have been read, no less."

  ([port n]

   (read port
         n
         0))


  ([^SerialPort port n timeout-ms]

   (.setSerialPortTimeout port
                          timeout-ms)
   (try
     (if (= n
            1)
       (list (.read (input-stream port)))
       (let [ba (byte-array n)]
         (take (.read (input-stream port)
                 ba)
               (map (fn to-unsigned [b]
                      (bit-and b
                               0xff))
                    ba))))
      (catch SerialPortTimeoutException _
        '()))))


;;;;;;;;;; Writing to an output stream



(defprotocol IWritable

  "Anything that can be converted to a byte-array and then be written using a serial port."

  (to-byte-array [this]

    "Converts the given argument to a byte array."))



(extend-protocol IWritable


  (class (byte-array 0))

    (to-byte-array [this]
      this)


  Character

    (to-byte-array [this]
      (int this))


  clojure.lang.Seqable

    (to-byte-array [this]
      (byte-array this))


  Number

    (to-byte-array [this]
      (int this))


  String

    (to-byte-array [this]
      (.getBytes this)))



(defn write

  "Writes to the serial port anything satisfying the [[IWritable]] protocol.
  
   Are already available:

   - Byte arrays
   - Characters
   - Sequences of any sort (seqables)
   - Number converted to a single unsigned byte
   - Strings"

  [^SerialPort port writable]

  (.write (output-stream port)
          (to-byte-array writable)))


src/clotty/core.clj
(ns clotty.core

  "IO on serial ports by wrapping jRxTx.
  
   Cf. https://github.com/openmuc/jrxtx"

  {:author "Adam Helinski"}

  (:import (java.io InputStream
                    OutputStream)
           (org.openmuc.jrxtx DataBits
                              StopBits
                              Parity
                              FlowControl
                              SerialPort
                              SerialPortBuilder
                              SerialPortException
                              SerialPortTimeoutException)))




;;;;;;;;;; Private interop fns


(defn- -to-data-bits

  "Converts a number to a DataBits enum value."

  ^DataBits

  [^long data-bits]

  (condp identical?
         data-bits
    5 DataBits/DATABITS_5
    6 DataBits/DATABITS_6
    7 DataBits/DATABITS_7
    8 DataBits/DATABITS_8))




(defn- -from-data-bits

  "Converts a Databits enum value to an number."

  ^long

  [^DataBits data-bits]

  (condp identical?
         data-bits
    DataBits/DATABITS_5 5
    DataBits/DATABITS_6 6
    DataBits/DATABITS_7 7
    DataBits/DATABITS_8 8))




(defn- -to-stop-bits

  "Converts a number to a StopBits enum value."

  ^StopBits

  [^long stop-bits]

  (condp identical?
         stop-bits
    1   StopBits/STOPBITS_1
    1.5 StopBits/STOPBITS_1_5
    2   StopBits/STOPBITS_2))




(defn- -from-stop-bits

  "Converts a StopBits enum value to a number."

  [^StopBits stop-bits]

  (condp identical?
         stop-bits
    StopBits/STOPBITS_1   1
    StopBits/STOPBITS_1_5 1.5
    StopBits/STOPBITS_2   2))




(defn- -to-parity

  "Converts a kw to a Parity enum value."

  ^Parity

  [^clojure.lang.Keyword parity]

  (condp identical?
         parity
    :even  Parity/EVEN
    :mark  Parity/MARK
    :none  Parity/NONE
    :odd   Parity/ODD
    :space Parity/SPACE))




(defn- -from-parity

  "Converts a Parity enum value to a kw."

  ^clojure.lang.Keyword

  [^Parity parity]

  (condp identical?
         parity
    Parity/EVEN  :even
    Parity/MARK  :mark
    Parity/NONE  :none
    Parity/ODD   :odd
    Parity/SPACE :space))




(defn- -to-flow-control

  "Converts a kw to a FlowControl enum value."

  ^FlowControl

  [^clojure.lang.Keyword flow-control]

  (condp identical?
         flow-control
    :none     FlowControl/NONE
    :rts-cts  FlowControl/RTS_CTS
    :xon-xoff FlowControl/XON_XOFF))




(defn- -from-flow-control

  "Converts a FlowControl enum value to a kw."

  ^clojure.lang.Keyword

  [^FlowControl flow-control]

  (condp identical?
         flow-control
    FlowControl/NONE     :none
    FlowControl/RTS_CTS  :rts-cts
    FlowControl/XON_XOFF :xon-xoff))




;;;;;;;;;; Private helpers


(defmacro ^:private -read

  "Helper for `read-byte` and `read-bytes`.
  
   Catches timeout exceptions."

  [port timeout & body]

  `(do
     (.setSerialPortTimeout ~port
                            ~timeout)
     (try
       ~@body
       (catch SerialPortTimeoutException _#
         nil)
       (catch SerialPortException e#
         (if (.isClosed ~port)
           -1
           (throw e#))))))




(defn- -lz-seq

  "Gets a lazy sequence for reading a serial port byte per byte with a timeout."

  [^SerialPort port ^InputStream is timeout]

  (lazy-seq (let [b (-read port
                           timeout
                           (.read is))]
              (when (and (some? b)
                         (not= b
                               -1))
                (cons b
                      (-lz-seq port
                               is
                               timeout))))))




(defmacro ^:private -write

  "Helper for `write-byte` and `write-bytes`.
  
   Catches timeout exceptions."

  [port & body]

  `(try
     ~@body
     true
     (catch SerialPortException e#
       (if (closed? ~port)
         false
         (throw e#)))))




;;;;;;;;;;


(defprotocol IWrapper

  "Commands a serial port."

  (raw [this]

    "Gets the raw org.openmuc.jrxtx.SerialPort object.")


  (describe [this]

    "Describes the serial port as a map.")


  (configure [this config]

    "Re-configures the serial port (eg. change baud-rate) on the fly.

    
     Throws
    
       java.io.IOException
         If something goes wrong.


     Cf. `open`")


  (available-bytes [this]

    "Gets the number of bytes that can be read from the buffer before blocking.

     Throws
    
       java.io.IOException
         If something goes wrong.")


  (read-byte [this]
             [this timeout]

    "Reads a single byte.
    
     A timeout in milliseconds can be provided, 0 meaning forever.
    

     Returns
    
       nil
         If the timeout is elapsed.

       -1
         If the port is closed.

       byte
         Otherwise.


     Throws
     
       java.io.IOException
         If something goes wrong.")


  (read-bytes [this]
              [this ba]
              [this ba timeout]
              [this ba offset n]
              [this ba offset n timeout]

  "Reads several bytes from the given serial port and writes them a byte array.
  
   A timeout in milliseconds can be provided, 0 meaning forever.
  

   Returns
  
     nil
       If the timeout is elapsed.

     -1
       If the port is closed.

     number of read bytes
       If a byte array is provided.

     byte array with the number of available bytes
       If no byte array is provided.
  

   Throws

     java.io.IOException
       If something goes wrong.
  

   Cf. `available-bytes`")


  (skip [this n]
        [this n timeout]

    "Skips over and discards `n` bytes of data. Just as reads, this function will
     block if `n` is greater than the number of bytes available in the buffer.

     A timeout in milliseconds can be provided, 0 meaning forever.

    
     Returns

       nil
         If the timeout is elapsed.

       actual number of bytes skipped
         Otherwise.


     Throws
    
       java.io.IOException
         If something goes wrong.")


  (seq-timeout [this timeout]

    "Converts to a sequence for reading byte per byte with a timeout, 0 meaning forever.

     Cf. `read-byte`")


  (write-byte [this b]

    "Writes a single byte to the given port.

     Returns
    
       true
         If the write is succesful.
     
       false
         If the port is closed.


     Throws
     
      java.io.IOException
        If something goes wrong.")


  (write-bytes [this ba]
               [this ba offset n]

    "Writes a byte array to the given port.
    

     Returns
    
       true
         If the write is succesful.
     
       false
         If the port is closed.


     Throws
     
      java.io.IOException
        If something goes wrong.")


  (closed? [this]

    "Is the serial port closed?")


  (close [this]

    "Closes the serial port.
    
     Throws
    
       java.io.IOException
         If something goes wrong."))




(deftype Wrapper [^SerialPort   port
                  ^InputStream  is
                  ^OutputStream os]

  IWrapper


    (raw [_]
      port)


    (describe [_]
      {:port         (.getPortName                        port)
       :closed?      (.isClosed                           port)
       :baud-rate    (.getBaudRate                        port)
       :data-bits    (-from-data-bits    (.getDataBits    port))
       :stop-bits    (-from-stop-bits    (.getStopBits    port))
       :parity       (-from-parity       (.getParity      port))
       :flow-control (-from-flow-control (.getFlowControl port))})


    (configure [_ config]
      (doseq [[k
               v] config]
        (case k
          :baud-rate    (.setBaudRate    port
                                         v)
          :data-bits    (.setDataBits    port
                                         (-to-data-bits v))
          :stop-bits    (.setStopBits    port
                                         (-to-stop-bits v))
          :parity       (.setParity      port
                                         (-to-parity v))
          :flow-control (.setFlowControl port
                                         (-to-flow-control v))))
      port)


    (available-bytes [_]
      (.available is))


    (read-byte [_ timeout]
      (-read port
             timeout
             (.read is)))


    (read-byte [this]
      (read-byte this
                 0))


    (read-bytes [_ ba timeout]

     (-read port
            timeout
            (.read is
                   ba)))


    (read-bytes [this ba]

     (read-bytes this
                 ba
                 0))


    (read-bytes [this]

     (let [n (available-bytes port)]
       (when (pos? n)
         (let [ba (byte-array n)]
           (read-bytes this
                       ba
                       0)
           ba))))


    (read-bytes [_ ba offset n timeout]

     (-read port
            timeout
            (.read is
                   ba
                   offset
                   n)))


    (read-bytes [this ba offset n]

     (read-bytes this
                 ba
                 offset
                 n
                 0))


    (skip [_ n timeout]
      (-read port
             timeout
             (.skip is
                    n)))


    (skip [this n]
      (skip this
            n
            0))


    (seq-timeout [_ timeout]
      (-lz-seq port
               is
               timeout))


    (write-byte [_ n]
      (-write port
              (.write os
                      ^long n)))


    (write-bytes [_ ba]
      (-write port
              (.write os
                      ^bytes ba)))


    (write-bytes [_ ba offset n]
      (-write port
              (.write os
                      ba
                      offset
                      n)))


    (closed? [_]
      (.isClosed port))


    (close [_]
      (.close port))




  clojure.lang.Seqable

    (seq [_]
      (-lz-seq port
               is
               0)))




;;;;;;;;;;


(defn available-ports

  "Gets a set of available serial ports.

   Prints an error if a port is already locked but does not throw."

  []

  (into #{}
        (map str)
        (SerialPortBuilder/getSerialPortNames)))




(defn open

  "Opens a serial port.

   The configuration map may contain

      :baud-rate
          Preferably a standard baud rate

      :data-bits
          How many bits per data char
          #{5 6 7 8}

      :stop-bits
          Bits sent at the end of each data char
          #{1 1.5 2}

      :parity
          Error detection
          #{:even   ;; An even parity bit will be sent for each data char
            :mark   ;; Mark parity bit
            :none   ;; No parity bit
            :odd    ;; Odd parity will be sent for each data char
            :space  ;; Space parity
            }

      :flow-control
          Handshaking for preventing the sender from sending too much too fast
          #{:none      ;; No handshaking
            :rts-cts   ;; Hardware flow-control
            :xon-xoff  ;; Software flow-control
            }


   Ex. (open \"/dev/ttyUSB0\"
             {:baud-rate 2400
              :parity    :none})
   

   Throws

     java.io.IOException
       If something goes wrong.

     com.openmuc.jrxtx.PortNotFoundException
       If the port is busy or does not exist."

  ^Wrapper

  [path & [{:as   config
            :keys [baud-rate
                   data-bits
                   stop-bits
                   parity
                   flow-control]
            :or   {baud-rate    9600
                   data-bits    8
                   stop-bits    1
                   parity       :even
                   flow-control :none}}]]

  (let [raw ^SerialPort (-> (SerialPortBuilder/newBuilder path)
                            (.setBaudRate    baud-rate)
                            (.setDataBits    (-to-data-bits    data-bits))
                            (.setStopBits    (-to-stop-bits    stop-bits))
                            (.setParity      (-to-parity       parity))
                            (.setFlowControl (-to-flow-control flow-control))
                            .build)]
    (Wrapper. raw
              (.getInputStream raw)
              (.getOutputStream raw))))
